File: mac/Source_objc/BluetoothScanner.m
//
//  BluetoothScanner.m
//  Created by Michael Simms on 5/8/22.
//

#import "BluetoothScanner.h"

@implementation BluetoothScanner

#pragma mark Internal state management functions

- (void)startTrackingConnectedPeripheral:(CBPeripheral*)peripheral
{
	bool found = false;
	
	if (!self->connectedPeripherals)
	{
		self->connectedPeripherals = [[NSMutableArray alloc] init];
	}

	for (CBPeripheral* temp in self->connectedPeripherals)
	{
		if (temp == peripheral)
		{
			found = true;
			break;
		}
	}

	if (!found)
	{
		[self->connectedPeripherals addObject:peripheral];
		peripheral.delegate = self;
	}
}

- (void)stopTrackingConnectedPeripheral:(CBPeripheral*)peripheral
{
	[self->connectedPeripherals removeObject:peripheral];
}

- (bool)isTrackedPeripheral:(CBPeripheral*)peripheral
{
	for (CBPeripheral* temp in self->connectedPeripherals)
	{
		if (temp == peripheral)
		{
			return true;
		}
	}
	return false;
}

- (void)restart
{
	if ([self->centralManager state] == CBManagerStatePoweredOn)
	{
		NSDictionary* options = [NSDictionary dictionaryWithObject:[NSNumber numberWithBool:NO] forKey:CBCentralManagerScanOptionAllowDuplicatesKey];
		[self->centralManager scanForPeripheralsWithServices:self->serviceIdsToScanFor options:options];
	}
}

#pragma mark Central Manager callbacks

- (void)centralManager:(CBCentralManager*)central didConnectPeripheral:(CBPeripheral*)peripheral
{
	// Discover any relevant services.
	if ([self isTrackedPeripheral:peripheral])
	{
		[peripheral discoverServices:self->serviceIdsToScanFor];
	}
}

- (void)centralManager:(CBCentralManager*)central didDisconnectPeripheral:(CBPeripheral*)peripheral error:(NSError*)error
{
	// Remove the peripheral from the connected list.
	[self stopTrackingConnectedPeripheral:peripheral];
}

- (void)centralManager:(CBCentralManager*)central didDiscoverPeripheral:(CBPeripheral*)peripheral advertisementData:(NSDictionary*)advertisementData RSSI:(NSNumber*)RSSI
{
	// Notify the callback.
	// If the callback returns true then we should connect to the peripheral.
	if (self->peripheralDiscoveryCallback(peripheral, advertisementData.description, self->callbackParam))
	{
		[self->centralManager connectPeripheral:peripheral options:nil];
		[self startTrackingConnectedPeripheral:peripheral];
	}
}

- (void)centralManager:(CBCentralManager*)central didFailToConnectPeripheral:(CBPeripheral*)peripheral error:(NSError*)error
{
}

- (void)centralManager:(CBCentralManager*)central didRetrieveConnectedPeripherals:(NSArray*)peripherals
{
}

- (void)centralManager:(CBCentralManager*)central didRetrievePeripherals:(NSArray*)peripherals
{
}

- (void)centralManagerDidUpdateState:(CBCentralManager*)central
{
	[self restart];
}

#pragma mark Peripheral callbacks

- (void)peripheral:(CBPeripheral*)peripheral didDiscoverServices:(NSError*)error
{
	for (CBService* service in peripheral.services)
	{
		for (CBUUID* value in self->serviceIdsToScanFor)
		{
			if ([value isEqual:[service UUID]])
			{
				// Notify callbacks.
				self->serviceDiscoveryCallback(peripheral, [service UUID], self->callbackParam);
				
				// Discover characteristics.
				[peripheral discoverCharacteristics:nil forService:service];
			}
		}
	}
}

- (void)peripheral:(CBPeripheral*)peripheral didDiscoverIncludedServicesForService:(CBService *)service error:(NSError*)error
{
}

- (void)peripheral:(CBPeripheral*)peripheral didDiscoverCharacteristicsForService:(CBService*)service error:(NSError*)error
{
	for (CBCharacteristic* characteristic in service.characteristics)
	{
		[peripheral setNotifyValue:TRUE forCharacteristic:characteristic];
		[peripheral readValueForCharacteristic:characteristic];
	}
}

- (void)peripheral:(CBPeripheral*)peripheral didUpdateValueForCharacteristic:(CBCharacteristic*)characteristic error:(NSError*)error
{
	if (characteristic.value != nil)
	{
		self->valueUpdatedCallback(peripheral, characteristic.service.UUID, characteristic.UUID, characteristic.value, self->callbackParam);
	}
}

- (void)peripheral:(CBPeripheral*)peripheral didUpdateValueForDescriptor:(CBDescriptor*)descriptor error:(NSError*)error
{
}

- (void)peripheral:(CBPeripheral*)peripheral didWriteValueForCharacteristic:(CBCharacteristic*)characteristic error:(NSError*)error
{
}

- (void)peripheral:(CBPeripheral*)peripheral didWriteValueForDescriptor:(CBDescriptor*)descriptor error:(NSError*)error
{
}

- (void)peripheral:(CBPeripheral*)peripheral didUpdateNotificationStateForCharacteristic:(CBCharacteristic*)characteristic error:(NSError*)error
{
}

#pragma mark Public interface for this class

- (void)start:(NSArray*)serviceIdsToScanFor withPeripheralDiscoveredCallback:(peripheralDiscoveredCb)peripheralCallback withServiceCallback:(serviceEnumeratedCb)serviceCallback withValueUpdatedCallback:(valueUpdatedCb)valueUpdatedCallback withCallbackParam:(void*)callbackParam
{
	self->serviceIdsToScanFor = serviceIdsToScanFor;
	self->peripheralDiscoveryCallback = peripheralCallback;
	self->serviceDiscoveryCallback = serviceCallback;
	self->valueUpdatedCallback = valueUpdatedCallback;
	self->callbackParam = callbackParam;
	self->centralManager = [[CBCentralManager alloc] initWithDelegate:self queue:nil];
}

- (void)stop
{
	self->centralManager = nil;
}

@end

File: mac/BluetoothTest/Shared/BluetoothScanner.swift
//
//  BluetoothScanner.swift
//  BluetoothTest
//
//  Created by Michael Simms on 5/3/22.
//

import Foundation
import CoreBluetooth

class BluetoothScanner: NSObject, CBPeripheralDelegate, CBCentralManagerDelegate, ObservableObject {
	
	@Published var connectedPeripherals: Array<CBPeripheral> = []

	private var centralManager: CBCentralManager!
	private var serviceIdsToScanFor: Array<CBUUID> = []
	private var peripheralDiscoveryCallbacks: Array<(String) -> Bool> = []
	private var serviceDiscoveryCallbacks: Array<() -> Bool> = []
	private var valueUpdatedCallbacks: Array<(CBPeripheral, Data) -> Void> = []

	func startTrackingConnectedPeripheral(peripheral: CBPeripheral) {
		var found = false

		for temp in self.connectedPeripherals {
			if temp == peripheral {
				found = true
				break
			}
		}
	
		if !found {
			self.connectedPeripherals.append(peripheral)
			peripheral.delegate = self
		}
	}
	func stopTrackingConnectedPeripheral(peripheral: CBPeripheral) {
		for (index, value) in self.connectedPeripherals.enumerated() {
			if value == peripheral {
				self.connectedPeripherals.remove(at: index)
				break
			}
		}
	}
	func isTrackedPeripheral(peripheral: CBPeripheral) -> Bool {
		for temp in self.connectedPeripherals {
			if temp == peripheral {
				return true
			}
		}
		return false
	}
	
	///
	/// Central Manager callbacks
	///

	func centralManagerDidUpdateState(_ central: CBCentralManager) {
		if central.state == .poweredOn {
			centralManager.scanForPeripherals(withServices: self.serviceIdsToScanFor, options: [CBCentralManagerScanOptionAllowDuplicatesKey: true])
		}
	}
	func centralManager(_ central: CBCentralManager, willRestoreState dict: [String : Any]) {
	}
	func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String : Any], rssi RSSI: NSNumber) {

		// Notify callbacks
		for cb in self.peripheralDiscoveryCallbacks {

			// If the callback returns true then we should connect to the peripheral.
			if cb(advertisementData.description) {
				self.centralManager.connect(peripheral, options: nil)
				self.startTrackingConnectedPeripheral(peripheral: peripheral)
			}
		}
	}
	func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {

		// Discover any relevant services.
		if self.isTrackedPeripheral(peripheral: peripheral) {
			peripheral.discoverServices(self.serviceIdsToScanFor)
		}
	}
	func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
	}
	func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
		
		// Remove the peripheral from the connected list.
		self.stopTrackingConnectedPeripheral(peripheral: peripheral)
	}

	///
	/// Peripheral callbacks
	///

	func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
		if let services = peripheral.services {
			for service in services {
				for (_, value) in self.serviceIdsToScanFor.enumerated() {
					if value == service.uuid {
						peripheral.discoverCharacteristics(nil, for: service)
						return
					}
				}
			}
		}
	}
	func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
		if let characteristics = service.characteristics {
			for characteristic in characteristics {
				peripheral.setNotifyValue(true, for: characteristic)
				peripheral.readValue(for: characteristic)
			}
		}
	}
	func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
		for cb in self.valueUpdatedCallbacks {
			if characteristic.value != nil {
				cb(peripheral, characteristic.value!)
			}
		}
	}
	func peripheral(_ peripheral: CBPeripheral, didWriteValueFor characteristic: CBCharacteristic, error: Error?) {
	}

	func startScanning(serviceIdsToScanFor: Array<CBUUID>, peripheralCallbacks: Array<(String) -> Bool>, serviceCallbacks: Array<() -> Bool>, valueUpdatedCallbacks: Array<(CBPeripheral, Data) -> Void>) {
		self.serviceIdsToScanFor = serviceIdsToScanFor
		self.peripheralDiscoveryCallbacks = peripheralCallbacks
		self.serviceDiscoveryCallbacks = serviceCallbacks
		self.valueUpdatedCallbacks = valueUpdatedCallbacks
		self.centralManager = CBCentralManager(delegate: self, queue: nil)
	}
	func stopScanning() {
		self.centralManager = nil
	}
}
